#include "generalinfowidget.h"
#include "ui_generalinfowidget.h"

GeneralInfoWidget::GeneralInfoWidget(QWidget *parent)
    : QWidget(parent)
    , ui(new Ui::GeneralInfoWidget)
{
    ui->setupUi(this);

    // 1. –ù–∞–ª–∞—à—Ç–æ–≤—É—î–º–æ –≤–∏–≥–ª—è–¥ (—Å—Ç–∏–ª—ñ, –∫—É—Ä—Å–æ—Ä–∏, –ø–æ—á–∞—Ç–∫–æ–≤–∏–π —Å—Ç–∞–Ω)
    setupUI();

    // 2. –ü—ñ–¥–∫–ª—é—á–∞—î–º–æ —Å–∏–≥–Ω–∞–ª–∏ (–≤–∑–∞—î–º–æ–¥—ñ—è)
    createConnections();
}

GeneralInfoWidget::~GeneralInfoWidget()
{
    delete ui;
}

// --- –Ü–ù–Ü–¶–Ü–ê–õ–Ü–ó–ê–¶–Ü–Ø ---

void GeneralInfoWidget::setupUI()
{
    // –•–æ–≤–∞—î–º–æ —Ç–µ–ª–µ—Ñ–æ–Ω –∑–∞ –∑–∞–º–æ–≤—á—É–≤–∞–Ω–Ω—è–º
    ui->labelPhone->setVisible(false);

    // –ù–∞–ª–∞—à—Ç—É–≤–∞–Ω–Ω—è –∫–æ–Ω—Ç–µ–π–Ω–µ—Ä–∞ frameTitle
    // –í–º–∏–∫–∞—î–º–æ —Ä–µ–∂–∏–º CustomContextMenu, —â–æ–± —Ñ—Ä–µ–π–º —Ä–µ–∞–≥—É–≤–∞–≤ –Ω–∞ –ü–ö–ú
    ui->frameTitle->setContextMenuPolicy(Qt::CustomContextMenu);

    // –í—ñ–∑—É–∞–ª—å–Ω—ñ –ø—ñ–¥–∫–∞–∑–∫–∏
    ui->frameTitle->setCursor(Qt::PointingHandCursor);
    ui->frameTitle->setToolTip("–ù–∞—Ç–∏—Å–Ω—ñ—Ç—å –ø—Ä–∞–≤—É –∫–Ω–æ–ø–∫—É –º–∏—à—ñ –¥–ª—è –æ–ø—Ü—ñ–π");

    // –°—Ç–∏–ª—ñ–∑–∞—Ü—ñ—è: –ø—Ä–æ–∑–æ—Ä–∞ —Ä–∞–º–∫–∞, —è–∫–∞ —Å—Ç–∞—î –≤–∏–¥–∏–º–æ—é –ø—Ä–∏ –Ω–∞–≤–µ–¥–µ–Ω–Ω—ñ
    ui->frameTitle->setStyleSheet(R"(
        QFrame#frameTitle {
            border: 1px solid transparent;
            border-radius: 6px;
            padding: 5px;
        }
        QFrame#frameTitle:hover {
            background-color: #f1f3f4; /* –°–≤—ñ—Ç–ª–æ-—Å—ñ—Ä–∏–π */
            border: 1px solid #dadce0; /* –†–∞–º–∫–∞ */
        }
    )");
}

void GeneralInfoWidget::createConnections()
{
    // –í—Å—ñ connect –ø–∏—à–µ–º–æ –¢–Ü–õ–¨–ö–ò –¢–£–¢

    // –û–±—Ä–æ–±–∫–∞ –ø—Ä–∞–≤–æ–≥–æ –∫–ª—ñ–∫—É –Ω–∞ —Ñ—Ä–µ–π–º—ñ
    connect(ui->frameTitle, &QFrame::customContextMenuRequested,
            this, &GeneralInfoWidget::onFrameContextMenu);

    // --- –ù–ê–í–Ü–ì–ê–¶–Ü–Ø QStackedWidget ---
    // –ü–µ—Ä–µ–º–∏–∫–∞—î–º–æ —Å—Ç–æ—Ä—ñ–Ω–∫–∏ –ø—Ä–∏ –Ω–∞—Ç–∏—Å–∫–∞–Ω–Ω—ñ –Ω–∞ –≤—ñ–¥–ø–æ–≤—ñ–¥–Ω—É –≤–∫–ª–∞–¥–∫—É-–∫–Ω–æ–ø–∫—É
    connect(ui->toolButtonRRO, &QToolButton::clicked, this, [this](){
        ui->stackedWidgetInfo->setCurrentWidget(ui->pageRRO);
    });
    connect(ui->toolButtonPRK, &QToolButton::clicked, this, [this](){
        ui->stackedWidgetInfo->setCurrentWidget(ui->pagePRK);
    });
    connect(ui->toolButtonTanks, &QToolButton::clicked, this, [this](){
        ui->stackedWidgetInfo->setCurrentWidget(ui->pageTanks);
    });

    // --- –ö–ù–û–ü–ö–ê –ö–û–ü–Ü–Æ–í–ê–¢–ò –í–°–Ü –†–†–û ---
    connect(ui->toolButtonCopyAllRRO, &QToolButton::clicked, this, &GeneralInfoWidget::onCopyAllRROClicked);
}

// --- –õ–û–ì–Ü–ö–ê –î–ê–ù–ò–• ---

void GeneralInfoWidget::updateData(const StationDataContext::GeneralInfo &info)
{
    m_lastInfo = info;

    // --- –í–ò–ü–†–ê–í–õ–ï–ù–ù–Ø –°–¢–ê–¢–£–°–£ (CSS –∑–∞–º—ñ—Å—Ç—å Emoji) ---

    // 1. –û—á–∏—â–∞—î–º–æ —Ç–µ–∫—Å—Ç (—â–æ–± –Ω–µ –±—É–ª–æ –∫–≤–∞–¥—Ä–∞—Ç—ñ–≤)
    ui->labelStatusIcon->setText("");

    // 2. –§—ñ–∫—Å—É—î–º–æ —Ä–æ–∑–º—ñ—Ä, —â–æ–± —Ü–µ –±—É–≤ —ñ–¥–µ–∞–ª—å–Ω–∏–π –∫–≤–∞–¥—Ä–∞—Ç
    ui->labelStatusIcon->setFixedSize(16, 16);

    // 3. –í–∏–∑–Ω–∞—á–∞—î–º–æ –∫–æ–ª—ñ—Ä
    QString color;
    QString toolTip;

    if (info.isActive && info.isWork) {
        color = "#34A853"; // –ó–µ–ª–µ–Ω–∏–π
        toolTip = "–ê–∫—Ç–∏–≤–Ω–∞, –ü—Ä–∞—Ü—é—î";
    } else if (info.isActive && !info.isWork) {
        color = "#FBBC05"; // –ñ–æ–≤—Ç–∏–π
        toolTip = "–ê–∫—Ç–∏–≤–Ω–∞, –ù–ï –ø—Ä–∞—Ü—é—î";
    } else {
        color = "#BDBDBD"; // –°—ñ—Ä–∏–π
        toolTip = "–ù–ï –ê–∫—Ç–∏–≤–Ω–∞";
    }

    // 4. –ú–∞–ª—é—î–º–æ –∫—Ä—É–∂–µ—á–æ–∫ —á–µ—Ä–µ–∑ Style Sheet
    // border-radius: 8px (—Ü–µ –ø–æ–ª–æ–≤–∏–Ω–∞ –≤—ñ–¥ —Ä–æ–∑–º—ñ—Ä—É 16px) —Ä–æ–±–∏—Ç—å –∫–æ–ª–æ
    ui->labelStatusIcon->setStyleSheet(QString(
                                           "QLabel { "
                                           "  background-color: %1; "
                                           "  border-radius: 8px; "
                                           "  border: none; "
                                           "}"
                                           ).arg(color));

    ui->labelStatusIcon->setToolTip(toolTip);

    // --- –†–ï–®–¢–ê –ö–û–î–£ –ë–ï–ó –ó–ú–Ü–ù ---
    ui->labelTitle->setText(QString("%1, –ê–ó–° %2").arg(info.clientName).arg(info.terminalId));
    ui->labelAddress->setText(QString("üìç %1").arg(info.address));

    if (!info.phone.isEmpty()) {
        ui->labelPhone->setText(QString("üìû %1").arg(info.phone));
        ui->labelPhone->setVisible(true);
    } else {
        ui->labelPhone->setVisible(false);
    }
}

// --- –õ–û–ì–Ü–ö–ê –ú–ï–ù–Æ ---

void GeneralInfoWidget::onFrameContextMenu(const QPoint &pos)
{
    QMenu menu(this);
    // –°—Ç–∏–ª—å –º–µ–Ω—é (–æ–ø—Ü—ñ–æ–Ω–∞–ª—å–Ω–æ)
    menu.setStyleSheet("QMenu { background: white; border: 1px solid #ccc; }"
                       "QMenu::item { padding: 5px 20px; }"
                       "QMenu::item:selected { background: #e8f0fe; color: black; }");

    // 1. –î—ñ—è: –ö–æ–ø—ñ—é–≤–∞—Ç–∏ –≤—Å–µ
    QAction *actCopyAll = menu.addAction("üìã –ö–æ–ø—ñ—é–≤–∞—Ç–∏ –≤—Å–µ");
    connect(actCopyAll, &QAction::triggered, this, [this](){
        copyToClipboard(false);
    });

    // 2. –î—ñ—è: –ö–æ–ø—ñ—é–≤–∞—Ç–∏ —Ç–µ–ª–µ—Ñ–æ–Ω (—Ç—ñ–ª—å–∫–∏ —è–∫—â–æ —î)
    if (!m_lastInfo.phone.isEmpty()) {
        QAction *actCopyPhone = menu.addAction("üìû –ö–æ–ø—ñ—é–≤–∞—Ç–∏ —Ç–µ–ª–µ—Ñ–æ–Ω");
        connect(actCopyPhone, &QAction::triggered, this, [this](){
            copyToClipboard(true);
        });
    }

    // 3. –î—ñ—è: –í—ñ–¥–∫—Ä–∏—Ç–∏ –∫–∞—Ä—Ç—É (—Ç—ñ–ª—å–∫–∏ —è–∫—â–æ —î –∫–æ–æ—Ä–¥–∏–Ω–∞—Ç–∏)
    if (m_lastInfo.latitude != 0.0 || m_lastInfo.longitude != 0.0) {
        menu.addSeparator();
        QAction *actOpenMap = menu.addAction("üó∫ –í—ñ–¥–∫—Ä–∏—Ç–∏ –∫–∞—Ä—Ç—É –≤ –±—Ä–∞—É–∑–µ—Ä—ñ");
        connect(actOpenMap, &QAction::triggered, this, &GeneralInfoWidget::openMapInBrowser);
    }

    // –í—ñ–¥–æ–±—Ä–∞–∂–∞—î–º–æ –º–µ–Ω—é –≤ —Ç–æ—á—Ü—ñ –∫–ª—ñ–∫—É
    menu.exec(ui->frameTitle->mapToGlobal(pos));
}

void GeneralInfoWidget::copyToClipboard(bool phoneOnly)
{
    QClipboard *clipboard = QApplication::clipboard();

    if (phoneOnly) {
        clipboard->setText(m_lastInfo.phone);
    } else {
        QStringList lines;
        lines << QString("%1, –ê–ó–° %2").arg(m_lastInfo.clientName).arg(m_lastInfo.terminalId);
        lines << m_lastInfo.address;

        if (!m_lastInfo.phone.isEmpty()) {
            lines << m_lastInfo.phone;
        }

        if (m_lastInfo.latitude != 0.0 || m_lastInfo.longitude != 0.0) {
            // –í–ò–ü–†–ê–í–õ–ï–ù–û: –í–∏–∫–æ—Ä–∏—Å—Ç–æ–≤—É—î–º–æ %1 —Ç–∞ %2 –¥–ª—è –ø—ñ–¥—Å—Ç–∞–Ω–æ–≤–∫–∏
            // –¢–∞–∫–æ–∂ –≤–∏–∫–æ—Ä–∏—Å—Ç–æ–≤—É—î–º–æ HTTPS —ñ —Å—Ç–∞–Ω–¥–∞—Ä—Ç–Ω–∏–π –¥–æ–º–µ–Ω
            QString link = QString("–ê–ó–° –Ω–∞ –º–∞–ø—ñ: https://www.google.com/maps?q=%1,%2")
                               .arg(m_lastInfo.latitude)
                               .arg(m_lastInfo.longitude);
            lines << link;
        }
        clipboard->setText(lines.join("\n"));
    }
}

void GeneralInfoWidget::openMapInBrowser()
{
    // –í–ò–ü–†–ê–í–õ–ï–ù–û: –ü—Ä–∞–≤–∏–ª—å–Ω–∏–π —Ñ–æ—Ä–º–∞—Ç URL –¥–ª—è –±—Ä–∞—É–∑–µ—Ä–∞
    QString urlStr = QString("https://www.google.com/maps?q=%1,%2")
                         .arg(m_lastInfo.latitude)
                         .arg(m_lastInfo.longitude);

    QDesktopServices::openUrl(QUrl(urlStr));
}
void GeneralInfoWidget::updateRROData(const QJsonArray &rroArray)
{
    m_lastRroArray = rroArray;

    // 1. –û–Ω–æ–≤–ª—é—î–º–æ –ª—ñ—á–∏–ª—å–Ω–∏–∫ —É —à–∞–ø—Ü—ñ
    ui->labelTotalRRO->setText(QString("üßæ –ó–Ω–∞–π–¥–µ–Ω–æ –∫–∞—Å–æ–≤–∏—Ö –∞–ø–∞—Ä–∞—Ç—ñ–≤: %1").arg(rroArray.size()));

    // 2. –û—á–∏—â–∞—î–º–æ —Å—Ç–∞—Ä—ñ –∫–∞—Ä—Ç–∫–∏ (—â–æ–± –≤–æ–Ω–∏ –Ω–µ –Ω–∞–∫–æ–ø–∏—á—É–≤–∞–ª–∏—Å—è –ø—Ä–∏ –∑–º—ñ–Ω—ñ –ê–ó–°)
    // –í–∞–∂–ª–∏–≤–æ: –º–∏ –Ω–µ –≤–∏–¥–∞–ª—è—î–º–æ –æ—Å—Ç–∞–Ω–Ω—ñ–π –µ–ª–µ–º–µ–Ω—Ç, –±–æ —Ü–µ –Ω–∞—à–∞ —Ä–æ–∑–ø—ñ—Ä–∫–∞ (QSpacerItem)!
    QLayoutItem *child;
    while (ui->layoutRROCards->count() > 1) {
        // –ó–∞–±–∏—Ä–∞—î–º–æ –Ω–∞–π–ø–µ—Ä—à–∏–π –µ–ª–µ–º–µ–Ω—Ç (–Ω—É–ª—å–æ–≤–∏–π)
        child = ui->layoutRROCards->takeAt(0);
        if (QWidget *widget = child->widget()) {
            widget->deleteLater();
        }
        delete child;
    }

    // 3. –î–æ–¥–∞—î–º–æ –Ω–æ–≤—ñ –∫–∞—Ä—Ç–∫–∏
    for (int i = 0; i < rroArray.size(); ++i) {
        QJsonObject jsonObj = rroArray[i].toObject();

        PosCardWidget *card = new PosCardWidget(this);
        // –ü–µ—Ä–µ–¥–∞—î–º–æ json, –ù–∞–∑–≤—É –∫–ª—ñ—î–Ω—Ç–∞ —Ç–∞ ID –ê–ó–° (–¥–ª—è –∫–æ–ø—ñ—é–≤–∞–Ω–Ω—è)
        card->setData(jsonObj, m_lastInfo.clientName, QString::number(m_lastInfo.terminalId));

        // –í—Å—Ç–∞–≤–ª—è—î–º–æ –∫–∞—Ä—Ç–∫—É –ü–ï–†–ï–î —Ä–æ–∑–ø—ñ—Ä–∫–æ—é
        ui->layoutRROCards->insertWidget(ui->layoutRROCards->count() - 1, card);
    }
}

// --- –ù–û–í–ò–ô –ú–ï–¢–û–î: –ö–û–ü–Ü–Æ–í–ê–ù–ù–Ø –í–°–¨–û–ì–û –°–ü–ò–°–ö–£ ---
void GeneralInfoWidget::onCopyAllRROClicked()
{
    if (m_lastRroArray.isEmpty()) return;

    QStringList lines;

    // 1. –®–∞–ø–∫–∞ –∑ –Ω–∞–∑–≤–æ—é –ê–ó–°
    lines << QString("üè¢ %1, –ê–ó–° %2").arg(m_lastInfo.clientName).arg(m_lastInfo.terminalId);
    lines << QString("üßæ –í—Å—å–æ–≥–æ –†–†–û: %1\n").arg(m_lastRroArray.size());

    // 2. –ü–µ—Ä–µ–±–∏—Ä–∞—î–º–æ –º–∞—Å–∏–≤ —ñ —Ñ–æ—Ä–º—É—î–º–æ —Ç–µ–∫—Å—Ç –¥–ª—è –∫–æ–∂–Ω–æ—ó –∫–∞—Å–∏
    for (int i = 0; i < m_lastRroArray.size(); ++i) {
        QJsonObject obj = m_lastRroArray[i].toObject();

        int posId = obj["pos_id"].toInt();
        QString manufacturer = obj["manufacturer"].toString();
        QString model = obj["model"].toString();
        QString zn = obj["factory_number"].toString();
        QString fn = obj["tax_number"].toString();
        QString softVer = obj["version"].toString();
        QString mukVer = obj["muk_version"].toString();
        QString regDate = obj["reg_date"].toString();

        lines << QString("üßæ –ö–∞—Å–∞ ‚Ññ%1 %2-%3").arg(posId).arg(manufacturer, model);
        lines << QString("   –ó–ù: %1").arg(zn);
        lines << QString("   –§–ù: %1").arg(fn);
        lines << QString("   üíª –ü–û: %1 | –ú–£–ö: %2").arg(softVer, mukVer);

        if (!regDate.isEmpty()) {
            lines << QString("   üìÖ –†–µ—î—Å—Ç—Ä–∞—Ü—ñ—è: %1").arg(regDate);
        }
        lines << ""; // –ü–æ—Ä–æ–∂–Ω—ñ–π —Ä—è–¥–æ–∫ –º—ñ–∂ –∫–∞—Å–∞–º–∏ –¥–ª—è –∫—Ä–∞—Å–∏
    }

    // 3. –í—ñ–¥–ø—Ä–∞–≤–ª—è—î–º–æ –≤ –±—É—Ñ–µ—Ä
    QClipboard *clipboard = QApplication::clipboard();
    clipboard->setText(lines.join("\n"));
}
